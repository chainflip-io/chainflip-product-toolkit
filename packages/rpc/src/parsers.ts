import { isNotNullish } from '@chainflip/utils/guard';
import { isHex } from '@chainflip/utils/string';
import { z } from 'zod';

// base types

export const hexString = z.string().refine(isHex, { message: 'Invalid hex string' });

export const u256 = hexString.transform((value) => BigInt(value));

export const numberOrHex = z.union([z.number().transform((n) => BigInt(n)), u256]);

const chainAssetMapFactory = <Z extends z.ZodTypeAny>(parser: Z, defaultValue: z.input<Z>) =>
  z.object({
    Bitcoin: z.object({ BTC: parser }),
    Ethereum: z.object({ ETH: parser, USDC: parser, FLIP: parser, USDT: parser }),
    Polkadot: z.object({ DOT: parser }),
    Arbitrum: z
      .object({ ETH: parser.default(defaultValue), USDC: parser.default(defaultValue) })
      .default({ ETH: defaultValue, USDC: defaultValue }),
    // Solana: z.object({ SOL: parser.default(defaultValue) }),
  });

const chainBaseAssetMapFactory = <Z extends z.ZodTypeAny>(parser: Z, defaultValue: z.input<Z>) =>
  z.object({
    Bitcoin: z.object({ BTC: parser }),
    Ethereum: z.object({ ETH: parser, FLIP: parser, USDT: parser }),
    Polkadot: z.object({ DOT: parser }),
    Arbitrum: z
      .object({ ETH: parser.default(defaultValue), USDC: parser.default(defaultValue) })
      .default({ ETH: defaultValue, USDC: defaultValue }),
    // Solana: z.object({ SOL: parser.default(defaultValue) }),
  });

const chainMapFactory = <Z extends z.ZodTypeAny>(parser: Z, defaultValue: z.input<Z>) =>
  z.object({
    Bitcoin: parser,
    Ethereum: parser,
    Polkadot: parser,
    Arbitrum: parser.default(defaultValue),
    // Solana: parser.default(defaultValue),
  });

const rpcAssetSchema = z.union([
  z.object({ chain: z.literal('Bitcoin'), asset: z.literal('BTC') }),
  z.object({ chain: z.literal('Polkadot'), asset: z.literal('DOT') }),
  z.object({ chain: z.literal('Ethereum'), asset: z.literal('FLIP') }),
  z.object({ chain: z.literal('Ethereum'), asset: z.literal('ETH') }),
  z.object({ chain: z.literal('Ethereum'), asset: z.literal('USDC') }),
  z.object({ chain: z.literal('Ethereum'), asset: z.literal('USDT') }),
  z.object({ chain: z.literal('Arbitrum'), asset: z.literal('ETH') }),
  z.object({ chain: z.literal('Arbitrum'), asset: z.literal('USDC') }),
]);

export type AssetAndChain = z.output<typeof rpcAssetSchema>;

type Rename<T, U extends Record<string, string>> = Omit<T, keyof U> & {
  [K in keyof U as NonNullable<U[K]>]: K extends keyof T ? T[K] : never;
};

const rename =
  <const U extends Record<string, string>>(mapping: U) =>
  <T>(obj: T): Rename<T, U> =>
    Object.fromEntries(
      Object.entries(obj as Record<string, unknown>).map(([key, value]) => [
        key in mapping ? mapping[key] : key,
        value,
      ]),
    ) as Rename<T, U>;

const rpcBaseResponse = z.object({
  id: z.union([z.string(), z.number()]),
  jsonrpc: z.literal('2.0'),
});

const nonNullish = z.any().refine(isNotNullish, { message: 'Value must not be null or undefined' });

const rpcSuccessResponse = rpcBaseResponse.extend({ result: nonNullish });

const rpcErrorResponse = rpcBaseResponse.extend({
  error: z.object({ code: z.number(), message: z.string() }),
});

export const rpcResponse = z.union([rpcSuccessResponse, rpcErrorResponse]);

// rpc responses

export const cfSwapRate = z.object({
  intermediary: numberOrHex.nullable(),
  output: numberOrHex,
});

const fee = z.intersection(rpcAssetSchema, z.object({ amount: numberOrHex }));

export const cfSwapRateV2 = z.object({
  egress_fee: fee,
  ingress_fee: fee,
  intermediary: u256.nullable(),
  network_fee: fee,
  output: u256,
});

export const chainGetBlockHash = hexString;

export const stateGetMetadata = hexString;

export const stateGetRuntimeVersion = z.object({
  specName: z.string(),
  implName: z.string(),
  authoringVersion: z.number(),
  specVersion: z.number(),
  implVersion: z.number(),
  apis: z.array(z.tuple([hexString, z.number()])),
  transactionVersion: z.number(),
  stateVersion: z.number(),
});

export const cfIngressEgressEnvironment = z
  .object({
    minimum_deposit_amounts: chainAssetMapFactory(numberOrHex, 0),
    ingress_fees: chainAssetMapFactory(numberOrHex.nullable(), null),
    egress_fees: chainAssetMapFactory(numberOrHex.nullable(), null),
    witness_safety_margins: chainMapFactory(z.number().nullable(), null),
    egress_dust_limits: chainAssetMapFactory(numberOrHex, 0),
    channel_opening_fees: chainMapFactory(numberOrHex, 0),
  })
  .transform(rename({ egress_dust_limits: 'minimum_egress_amounts' }));

export const cfSwappingEnvironment = z.object({
  maximum_swap_amounts: chainAssetMapFactory(numberOrHex.nullable(), null),
  network_fee_hundredth_pips: z.number(),
});

export const cfFundingEnvironment = z.object({
  redemption_tax: numberOrHex,
  minimum_funding_amount: numberOrHex,
});

export const cfPoolsEnvironment = z.object({
  fees: chainBaseAssetMapFactory(
    z.object({
      limit_order_fee_hundredth_pips: z.number(),
      range_order_fee_hundredth_pips: z.number(),
      range_order_total_fees_earned: z.object({ base: u256, quote: u256 }),
      limit_order_total_fees_earned: z.object({ base: u256, quote: u256 }),
      range_total_swap_inputs: z.object({ base: u256, quote: u256 }),
      limit_total_swap_inputs: z.object({ base: u256, quote: u256 }),
      quote_asset: z.object({ chain: z.literal('Ethereum'), asset: z.literal('USDC') }),
    }),
    {
      limit_order_fee_hundredth_pips: 0,
      range_order_fee_hundredth_pips: 0,
      range_order_total_fees_earned: { base: '0x0', quote: '0x0' },
      limit_order_total_fees_earned: { base: '0x0', quote: '0x0' },
      range_total_swap_inputs: { base: '0x0', quote: '0x0' },
      limit_total_swap_inputs: { base: '0x0', quote: '0x0' },
      quote_asset: { chain: 'Ethereum', asset: 'USDC' },
    },
  ),
});

export const cfEnvironment = z.object({
  ingress_egress: cfIngressEgressEnvironment,
  swapping: cfSwappingEnvironment,
  funding: cfFundingEnvironment,
  pools: cfPoolsEnvironment,
});

export const cfBoostPoolsDepth = z.array(
  z.intersection(rpcAssetSchema, z.object({ tier: z.number(), available_amount: u256 })),
);

export const cfSupportedAsssets = z
  .array(z.object({ chain: z.string(), asset: z.string() }))
  .transform((assets) =>
    assets.filter((asset): asset is AssetAndChain => rpcAssetSchema.safeParse(asset).success),
  );

export const brokerRequestSwapDepositAddress = z.object({
  address: z.string(),
  issued_block: z.number(),
  channel_id: z.number(),
  source_chain_expiry_block: numberOrHex,
  channel_opening_fee: u256,
});

export const unregistered = z.object({
  role: z.literal('unregistered'),
  flip_balance: numberOrHex,
});

export const broker = z.object({
  role: z.literal('broker'),
  flip_balance: numberOrHex,
  earned_fees: chainAssetMapFactory(numberOrHex, 0),
});

export const liquidityProvider = z.object({
  role: z.literal('liquidity_provider'),
  balances: chainAssetMapFactory(numberOrHex, '0x0'),
  refund_addresses: chainMapFactory(z.string().nullable(), null),
  flip_balance: numberOrHex,
  earned_fees: chainAssetMapFactory(numberOrHex, 0),
});

export const validator = z.object({
  role: z.literal('validator'),
  flip_balance: numberOrHex,
  bond: numberOrHex,
  last_heartbeat: z.number(),
  reputation_points: z.number(),
  keyholder_epochs: z.array(z.number()),
  is_current_authority: z.boolean(),
  is_current_backup: z.boolean(),
  is_qualified: z.boolean(),
  is_online: z.boolean(),
  is_bidding: z.boolean(),
  bound_redeem_address: hexString.nullable(),
  apy_bp: z.number().nullable(),
  restricted_balances: z.record(hexString, numberOrHex),
});

export const cfAccountInfo = z.union([unregistered, broker, liquidityProvider, validator]);

export const cfPoolPriceV2 = z.object({
  sell: numberOrHex.nullable(),
  buy: numberOrHex.nullable(),
  range_order: numberOrHex,
  base_asset: rpcAssetSchema,
  quote_asset: rpcAssetSchema,
});

const orderId = numberOrHex.transform((n) => String(n));

const limitOrder = z.object({
  id: orderId,
  tick: z.number(),
  sell_amount: numberOrHex,
  fees_earned: numberOrHex,
  original_sell_amount: numberOrHex,
  lp: z.string(),
});

const ask = limitOrder.transform((order) => ({
  ...order,
  type: 'ask' as const,
}));

const bid = limitOrder.transform((order) => ({
  ...order,
  type: 'bid' as const,
}));

export type RpcLimitOrder = z.infer<typeof ask> | z.infer<typeof bid>;

export type RpcRangeOrder = z.infer<typeof rangeOrder>;

const rangeOrder = z
  .object({
    id: orderId,
    range: z.object({ start: z.number(), end: z.number() }),
    liquidity: numberOrHex,
    fees_earned: z.object({ base: numberOrHex, quote: numberOrHex }),
    lp: z.string(),
  })
  .transform((order) => ({ ...order, type: 'range' as const }));

export const cfPoolOrders = z.object({
  limit_orders: z.object({
    asks: z.array(ask),
    bids: z.array(bid),
  }),
  range_orders: z.array(rangeOrder),
});

const boostPoolAmount = z.object({
  account_id: z.string(),
  amount: u256,
});
export const cfBoostPoolDetails = z.array(
  z.intersection(
    rpcAssetSchema,
    z.object({
      fee_tier: z.number(),
      available_amounts: z.array(boostPoolAmount),
      deposits_pending_finalization: z.array(
        z.object({
          deposit_id: z.number().transform(BigInt),
          owed_amounts: z.array(boostPoolAmount),
        }),
      ),
      pending_withdrawals: z.array(
        z.object({
          account_id: z.string(),
          pending_deposits: z.array(z.bigint()),
        }),
      ),
    }),
  ),
);

export const cfBoostPoolPendingFees = z.array(
  z.intersection(
    rpcAssetSchema,
    z.object({
      fee_tier: z.number(),
      pending_fees: z.array(
        z.object({
          deposit_id: z.number().transform(BigInt),
          fees: z.array(boostPoolAmount),
        }),
      ),
    }),
  ),
);
