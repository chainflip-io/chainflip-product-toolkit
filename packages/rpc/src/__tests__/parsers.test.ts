import { describe, expect, it } from 'vitest';
import { cfBoostPoolDetails, cfBoostPoolPendingFees, cfEnvironment, numberOrHex } from '../parsers';

describe('parsers', () => {
  describe('numberOrHex', () => {
    it.each([1, '0x1'])('transforms %s into a bigint', (value) => {
      expect(numberOrHex.parse(value)).toBe(BigInt(value));
    });
  });

  describe('cfBoostPoolDetails', () => {
    it('parses boost pool details', () => {
      cfBoostPoolDetails.parse([
        {
          fee_tier: 5,
          chain: 'Bitcoin',
          asset: 'BTC',
          available_amounts: [
            {
              account_id: 'cFL8fmgKZcchhtLagBH2GKfsuWxBqUaD5CYE1m7DFb8DBSLJ1',
              amount: '0xbed6146',
            },
          ],
          deposits_pending_finalization: [
            {
              deposit_id: 11,
              owed_amounts: [
                {
                  account_id: 'cFL8fmgKZcchhtLagBH2GKfsuWxBqUaD5CYE1m7DFb8DBSLJ1',
                  amount: '0xf4240',
                },
              ],
            },
          ],
          pending_withdrawals: [],
        },
        {
          fee_tier: 10,
          chain: 'Bitcoin',
          asset: 'BTC',
          available_amounts: [],
          deposits_pending_finalization: [],
          pending_withdrawals: [],
        },
      ]);
    });
  });

  describe('cfBoostPoolPendingFees', () => {
    it('parses boost pool pending fees', () => {
      cfBoostPoolPendingFees.parse([
        {
          fee_tier: 5,
          chain: 'Bitcoin',
          asset: 'BTC',
          pending_fees: [
            {
              deposit_id: 3,
              fees: [
                {
                  account_id: 'cFL8fmgKZcchhtLagBH2GKfsuWxBqUaD5CYE1m7DFb8DBSLJ1',
                  amount: '0x1f4',
                },
              ],
            },
            {
              deposit_id: 11,
              fees: [
                {
                  account_id: 'cFL8fmgKZcchhtLagBH2GKfsuWxBqUaD5CYE1m7DFb8DBSLJ1',
                  amount: '0x1f4',
                },
              ],
            },
          ],
        },
        {
          fee_tier: 10,
          chain: 'Bitcoin',
          asset: 'BTC',
          pending_fees: [],
        },
        {
          fee_tier: 30,
          chain: 'Bitcoin',
          asset: 'BTC',
          pending_fees: [],
        },
      ]);
    });
  });

  describe('cfEnvironment', () => {
    it('parses the cfEnvironment response', () => {
      cfEnvironment.parse({
        ingress_egress: {
          minimum_deposit_amounts: {
            Ethereum: {
              ETH: '0x0',
              FLIP: '0x0',
              USDC: '0x0',
              USDT: '0x0',
            },
            Polkadot: {
              DOT: '0x0',
            },
            Bitcoin: {
              BTC: '0x0',
            },
            Arbitrum: {
              ETH: '0x0',
              USDC: '0x0',
            },
            Solana: {
              SOL: '0x0',
            },
          },
          ingress_fees: {
            Ethereum: {
              ETH: '0x55730',
              FLIP: '0x2db7b25',
              USDC: '0x0',
              USDT: '0x0',
            },
            Polkadot: {
              DOT: '0xbc28f20',
            },
            Bitcoin: {
              BTC: '0x4f',
            },
            Arbitrum: {
              ETH: '0x574b457d400',
              USDC: '0x150e',
            },
            Solana: {
              SOL: '0x2f62d048',
            },
          },
          egress_fees: {
            Ethereum: {
              ETH: '0x77a10',
              FLIP: '0x336ea89',
              USDC: '0x0',
              USDT: '0x0',
            },
            Polkadot: {
              DOT: '0xbc4d910',
            },
            Bitcoin: {
              BTC: '0xb3',
            },
            Arbitrum: {
              ETH: '0x74645ca7000',
              USDC: '0x1765',
            },
            // Solana: {
            //   SOL: '0x4787c48',
            // },
          },
          witness_safety_margins: {
            Arbitrum: 1,
            Ethereum: 2,
            // Solana: 1,
            Bitcoin: 2,
            Polkadot: null,
          },
          egress_dust_limits: {
            Ethereum: {
              ETH: '0x1',
              FLIP: '0x1',
              USDC: '0x1',
              USDT: '0x1',
            },
            Polkadot: {
              DOT: '0x1',
            },
            Bitcoin: {
              BTC: '0x258',
            },
            Arbitrum: {
              ETH: '0x1',
              USDC: '0x1',
            },
            Solana: {
              SOL: '0x1',
            },
          },
          channel_opening_fees: {
            Ethereum: '0x0',
            Bitcoin: '0x0',
            Arbitrum: '0x0',
            Polkadot: '0x0',
            Solana: '0x0',
          },
          max_swap_retry_duration_blocks: {
            Ethereum: 5,
            Bitcoin: 2,
            Arbitrum: 30,
            Polkadot: 30,
            Solana: 10,
          },
        },
        swapping: {
          maximum_swap_amounts: {
            Ethereum: {
              ETH: null,
              FLIP: null,
              USDC: null,
              USDT: null,
            },
            Polkadot: {
              DOT: null,
            },
            Bitcoin: {
              BTC: null,
            },
            Arbitrum: {
              ETH: null,
              USDC: null,
            },
            Solana: {
              SOL: null,
            },
          },
          network_fee_hundredth_pips: 1000,
        },
        funding: {
          redemption_tax: '0x4563918244f40000',
          minimum_funding_amount: '0x8ac7230489e80000',
        },
        pools: {
          fees: {
            Ethereum: {
              ETH: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0x60df5e33d7020d',
                  quote: '0x84b113',
                },
                limit_order_total_fees_earned: {
                  base: '0x210317f6bb404',
                  quote: '0x0',
                },
                range_total_swap_inputs: {
                  base: '0x49e809af8ec5e4686d',
                  quote: '0x653bc61c72',
                },
                limit_total_swap_inputs: {
                  base: '0x192f8b3555c5aae8e',
                  quote: '0x0',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
              FLIP: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0x8fe2c5138911193',
                  quote: '0x1ad471',
                },
                limit_order_total_fees_earned: {
                  base: '0x0',
                  quote: '0x0',
                },
                range_total_swap_inputs: {
                  base: '0x6dc62318b26f5210b65',
                  quote: '0x1478117200',
                },
                limit_total_swap_inputs: {
                  base: '0x0',
                  quote: '0x0',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
              USDC: null,
              USDT: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0x8ae06',
                  quote: '0x1bf876',
                },
                limit_order_total_fees_earned: {
                  base: '0x0',
                  quote: '0x0',
                },
                range_total_swap_inputs: {
                  base: '0x69f36e6b2',
                  quote: '0x1556dc0dcb',
                },
                limit_total_swap_inputs: {
                  base: '0x0',
                  quote: '0x0',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
            },
            Polkadot: {
              DOT: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0x1264878c',
                  quote: '0x111676',
                },
                limit_order_total_fees_earned: {
                  base: '0x0',
                  quote: '0x0',
                },
                range_total_swap_inputs: {
                  base: '0xe084044b0fc',
                  quote: '0xd095d91a1',
                },
                limit_total_swap_inputs: {
                  base: '0x0',
                  quote: '0x0',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
            },
            Bitcoin: {
              BTC: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0x8961',
                  quote: '0xda0f22',
                },
                limit_order_total_fees_earned: {
                  base: '0x0',
                  quote: '0xac0c',
                },
                range_total_swap_inputs: {
                  base: '0x68cedbea',
                  quote: '0xa65ccd0506',
                },
                limit_total_swap_inputs: {
                  base: '0x0',
                  quote: '0x8341e381',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
            },
            Arbitrum: {
              ETH: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0x1a96a7c9901240',
                  quote: '0x210230',
                },
                limit_order_total_fees_earned: {
                  base: '0x0',
                  quote: '0x0',
                },
                range_total_swap_inputs: {
                  base: '0x1448f64c4ff15c61c0',
                  quote: '0x192ecd80a2',
                },
                limit_total_swap_inputs: {
                  base: '0x0',
                  quote: '0x0',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
              USDC: {
                limit_order_fee_hundredth_pips: 20,
                range_order_fee_hundredth_pips: 20,
                range_order_total_fees_earned: {
                  base: '0xaa9c1',
                  quote: '0x1b1d4b',
                },
                limit_order_total_fees_earned: {
                  base: '0x0',
                  quote: '0x0',
                },
                range_total_swap_inputs: {
                  base: '0x8228e1e3c',
                  quote: '0x14afa63498',
                },
                limit_total_swap_inputs: {
                  base: '0x0',
                  quote: '0x0',
                },
                quote_asset: {
                  chain: 'Ethereum',
                  asset: 'USDC',
                },
              },
            },
          },
        },
      });
    });
  });
});
