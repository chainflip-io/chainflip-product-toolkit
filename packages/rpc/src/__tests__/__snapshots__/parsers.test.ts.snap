// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parsers > cfAccountInfo > parses the latest cfAccountInfo schema for an lp 1`] = `
{
  "asset_balances": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 19842n,
    },
    "Ethereum": {
      "ETH": 0n,
      "FLIP": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 0n,
      "USDC": 0n,
    },
  },
  "balances": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 19842n,
    },
    "Ethereum": {
      "ETH": 0n,
      "FLIP": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 0n,
      "USDC": 0n,
    },
  },
  "bond": 0n,
  "boost_balances": {
    "Arbitrum": {
      "ETH": [],
      "USDC": [],
    },
    "Assethub": {
      "DOT": [],
      "USDC": [],
      "USDT": [],
    },
    "Bitcoin": {
      "BTC": [
        {
          "available_balance": 200000000n,
          "fee_tier": 5,
          "in_use_balance": 0n,
          "is_withdrawing": false,
          "total_balance": 200000000n,
        },
        {
          "available_balance": 200000000n,
          "fee_tier": 10,
          "in_use_balance": 0n,
          "is_withdrawing": false,
          "total_balance": 200000000n,
        },
        {
          "available_balance": 200000000n,
          "fee_tier": 30,
          "in_use_balance": 0n,
          "is_withdrawing": false,
          "total_balance": 200000000n,
        },
      ],
    },
    "Ethereum": {
      "ETH": [],
      "FLIP": [],
      "USDC": [],
      "USDT": [],
    },
    "Polkadot": {
      "DOT": [],
    },
    "Solana": {
      "SOL": [],
      "USDC": [],
    },
  },
  "collateral_balances": [],
  "earned_fees": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 0n,
    },
    "Ethereum": {
      "ETH": 0n,
      "FLIP": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 0n,
      "USDC": 0n,
    },
  },
  "estimated_redeemable_balance": 99999975540394000000n,
  "flip_balance": 99999975540394000000n,
  "lending_positions": [
    {
      "asset": "BTC",
      "available_amount": 200000000n,
      "chain": "Bitcoin",
      "total_amount": 200000000n,
    },
  ],
  "refund_addresses": {
    "Arbitrum": null,
    "Assethub": null,
    "Bitcoin": "mo3MtB6mLxTBBSmzzJvR3TtgrT9qBkoup3",
    "Ethereum": null,
    "Polkadot": null,
    "Solana": null,
  },
  "role": "liquidity_provider",
  "vanity_name": "Chainflip Testnet LP BOOST",
}
`;

exports[`parsers > cfAccountInfo > parses the latest cfAccountInfo schema for an lp on networks that haven't been upgraded yet 1`] = `
{
  "asset_balances": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 19842n,
    },
    "Ethereum": {
      "ETH": 0n,
      "FLIP": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 0n,
      "USDC": 0n,
    },
  },
  "balances": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 19842n,
    },
    "Ethereum": {
      "ETH": 0n,
      "FLIP": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 0n,
      "USDC": 0n,
    },
  },
  "bond": 0n,
  "boost_balances": {
    "Arbitrum": {
      "ETH": [],
      "USDC": [],
    },
    "Assethub": {
      "DOT": [],
      "USDC": [],
      "USDT": [],
    },
    "Bitcoin": {
      "BTC": [
        {
          "available_balance": 200000000n,
          "fee_tier": 5,
          "in_use_balance": 0n,
          "is_withdrawing": false,
          "total_balance": 200000000n,
        },
        {
          "available_balance": 200000000n,
          "fee_tier": 10,
          "in_use_balance": 0n,
          "is_withdrawing": false,
          "total_balance": 200000000n,
        },
        {
          "available_balance": 200000000n,
          "fee_tier": 30,
          "in_use_balance": 0n,
          "is_withdrawing": false,
          "total_balance": 200000000n,
        },
      ],
    },
    "Ethereum": {
      "ETH": [],
      "FLIP": [],
      "USDC": [],
      "USDT": [],
    },
    "Polkadot": {
      "DOT": [],
    },
    "Solana": {
      "SOL": [],
      "USDC": [],
    },
  },
  "collateral_balances": undefined,
  "earned_fees": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 0n,
    },
    "Ethereum": {
      "ETH": 0n,
      "FLIP": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 0n,
      "USDC": 0n,
    },
  },
  "estimated_redeemable_balance": 99999975540394000000n,
  "flip_balance": 99999975540394000000n,
  "lending_positions": undefined,
  "refund_addresses": {
    "Arbitrum": null,
    "Assethub": null,
    "Bitcoin": "mo3MtB6mLxTBBSmzzJvR3TtgrT9qBkoup3",
    "Ethereum": null,
    "Polkadot": null,
    "Solana": null,
  },
  "role": "liquidity_provider",
  "vanity_name": "Chainflip Testnet LP BOOST",
}
`;

exports[`parsers > cfAccountInfo > parses the old cfAccountInfo schema for an lp 1`] = `
{
  "balances": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 35910229n,
    },
    "Ethereum": {
      "ETH": 201081687243582975266n,
      "FLIP": 99996154262861338436158n,
      "USDC": 227879719689n,
      "USDT": 1999998411n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 3358768032n,
      "USDC": 0n,
    },
  },
  "boost_balances": {
    "Arbitrum": {
      "ETH": [],
      "USDC": [],
    },
    "Assethub": {
      "DOT": [],
      "USDC": [],
      "USDT": [],
    },
    "Bitcoin": {
      "BTC": [],
    },
    "Ethereum": {
      "ETH": [],
      "FLIP": [],
      "USDC": [],
      "USDT": [],
    },
    "Polkadot": {
      "DOT": [],
    },
    "Solana": {
      "SOL": [],
      "USDC": [],
    },
  },
  "earned_fees": {
    "Arbitrum": {
      "ETH": 0n,
      "USDC": 0n,
    },
    "Assethub": {
      "DOT": 0n,
      "USDC": 0n,
      "USDT": 0n,
    },
    "Bitcoin": {
      "BTC": 54632n,
    },
    "Ethereum": {
      "ETH": 18585729823626322n,
      "FLIP": 0n,
      "USDC": 718345809n,
      "USDT": 0n,
    },
    "Polkadot": {
      "DOT": 0n,
    },
    "Solana": {
      "SOL": 3358768032n,
      "USDC": 0n,
    },
  },
  "flip_balance": 9999712411635950000n,
  "refund_addresses": {
    "Arbitrum": null,
    "Assethub": null,
    "Bitcoin": "tb1qlpaajxzlpm6ds0jsg5dgsep56twkzs9ktsc8a9",
    "Ethereum": "0xe964a5775fd296ce1bd3e05547c59ce86ac077a2",
    "Polkadot": null,
    "Solana": "FkoiauqRvxTuimbohr2dK3UtwZ5Vak44dtSbK69nm1EY",
  },
  "role": "liquidity_provider",
}
`;
